### PR TITLE
OCPBUGS-17850: replace nonResourceURLs permissions by subresource

### PR DESCRIPTION
**Pull request by marioferh**
Don't allow common users access to restricted resources

### CODE REVIEW

File: jsonnet/components/cluster-monitoring-operator.libsonnet [hunk 1]
```diff
@@ -345,6 +347,43 @@ function(params) {
...
+  userWorkloadAlertmanagerApiReader: {
+    apiVersion: 'rbac.authorization.k8s.io/v1',
+    kind: 'Role',
+    metadata: {
+      name: 'monitoring-alertmanager-api-reader',
+      namespace: cfg.namespaceUserWorkload,
+    },
+    rules: [{
+      apiGroups: ['monitoring.coreos.com'],
+      resources: ['alertmanagers'],
```
Comment (simonpasquier) [inline jsonnet/components/cluster-monitoring-operator.libsonnet:364]: shouldn't it target the api subresource?
```suggestion
      resources: ['alertmanagers/api'],
```

Comment (marioferh) [reply jsonnet/components/cluster-monitoring-operator.libsonnet]: done

File: assets/prometheus-user-workload/alertmanager-user-workload-role-binding.yaml [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+    app.kubernetes.io/part-of: openshift-monitoring
+  name: alertmanager-user-workload-prometheususer-workload
+  namespace: openshift-user-workload-monitoring
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: Role
+  name: monitoring-alertmanager-api-writer
+subjects:
+- kind: ServiceAccount
+  name: user-workload
```
Comment (simonpasquier) [inline assets/prometheus-user-workload/alertmanager-user-workload-role-binding.yaml:15]: hmm the service account is rather `prometheus-user-workload`?

File: assets/prometheus-user-workload/alertmanager-user-workload-role-binding.yaml [hunk 3]
```diff
@@ -0,0 +1,16 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: RoleBinding
+metadata:
+  labels:
+    app.kubernetes.io/managed-by: cluster-monitoring-operator
+    app.kubernetes.io/part-of: openshift-monitoring
+  name: alertmanager-user-workload-prometheususer-workload
```
Comment (simonpasquier) [inline assets/prometheus-user-workload/alertmanager-user-workload-role-binding.yaml:7]: (nit)

```suggestion
  name: prometheus-user-workload-alertmanager-api
```

File: jsonnet/components/cluster-monitoring-operator.libsonnet [hunk 4]
```diff
@@ -345,6 +353,43 @@ function(params) {
     }],
   },
 
+  // This role provides read access to the user-workload Alertmanager API.
+  // We use a fake subresource 'api' to map to the /api/* endpoints of the
+  // Alertmanager API.
+  // Using "nonResourceURLs" doesn't work because authenticated users and
+  // service accounts are allowed to get /api/* by default.
```
Comment (simonpasquier) [inline jsonnet/components/cluster-monitoring-operator.libsonnet:360]:
```suggestion
  // service accounts are allowed to get /api/* by default.
  // See https://issues.redhat.com/browse/OCPBUGS-17850.
```

File: jsonnet/components/cluster-monitoring-operator.libsonnet [hunk 5]
```diff
@@ -358,6 +403,7 @@ function(params) {
     }],
   },
 
+  // This cluster role can be referenced in a RoleBinding object to provide read access to alerting and recording rules for a project.
```
Comment (simonpasquier) [inline jsonnet/components/cluster-monitoring-operator.libsonnet:406]: (nit)
```suggestion
  // This cluster role can be referenced in a RoleBinding object to provide read access to PrometheusRule objects for a project.
```

File: jsonnet/components/cluster-monitoring-operator.libsonnet [hunk 6]
```diff
@@ -399,6 +447,7 @@ function(params) {
     }],
   },
 
+  // This cluster role can be referenced in a RoleBinding object to provide read/write access to Alertmanager configurations for a project.
```
Comment (simonpasquier) [inline jsonnet/components/cluster-monitoring-operator.libsonnet:450]:
```suggestion
  // This cluster role can be referenced in a RoleBinding object to provide read/write access to AlertmanagerConfiguration objects for a project.
```

**Comment by juzhao** [comment]
tested with PR, 
1.  login with common user, with its token, common user can not view UWM alerts.
2. login with cluster admin user, with its token, cluster admin user can view UWM alerts, but prometheus-k8s/prometheus-user-workload SA can not view UWM alerts now, they had permission before
3. details see from bug comment

**Pull request merged**